Feed settings crash when rotating screen

First of all, thank you for this very useful Podcast app! While building the app myself from `develop`, I noticed an issue with the revised feed settings.

It appears the feed and its settings are not properly saved and restored in the `FeedSettingsActivity` (lifecycle methods). 

**App version**: unreleased/develop 1580e93b940436b76e2db46fa5501c8f02e38770

**Android version**: 7.1.2 (LineageOS 14.1-20180610-NIGHTLY-titan)

**Device model**: MOTO G 2014

**Expected behaviour**: View is restored after rotation

**Current behaviour**: App crashes

**First occurred**: On branch `develop` with revised feed settings

**Steps to reproduce**: in any feed, open settings, then rotate screen

**Environment**: Default settings on branch develop

**Stacktrace/Logcat**: 
```
2019-06-29 12:12:49.506 31847-31847/de.danoeh.antennapod.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: de.danoeh.antennapod.debug, PID: 31847
    java.lang.RuntimeException: Unable to start activity ComponentInfo{de.danoeh.antennapod.debug/de.danoeh.antennapod.activity.FeedSettingsActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'de.danoeh.antennapod.core.feed.FeedPreferences de.danoeh.antennapod.core.feed.Feed.getPreferences()' on a null object reference
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2684)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2751)
        at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4556)
        at android.app.ActivityThread.-wrap19(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1502)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6186)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'de.danoeh.antennapod.core.feed.FeedPreferences de.danoeh.antennapod.core.feed.Feed.getPreferences()' on a null object reference
        at de.danoeh.antennapod.fragment.FeedSettingsFragment.onCreatePreferences(FeedSettingsFragment.java:30)
        at android.support.v7.preference.PreferenceFragmentCompat.onCreate(PreferenceFragmentCompat.java:224)
        at android.support.v4.app.Fragment.performCreate(Fragment.java:2331)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1386)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1759)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1827)
        at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3244)
        at android.support.v4.app.FragmentManagerImpl.dispatchCreate(FragmentManager.java:3194)
        at android.support.v4.app.FragmentController.dispatchCreate(FragmentController.java:184)
        at android.support.v4.app.FragmentActivity.onCreate(FragmentActivity.java:355)
        at android.support.v7.app.AppCompatActivity.onCreate(AppCompatActivity.java:84)
        at de.danoeh.antennapod.activity.FeedSettingsActivity.onCreate(FeedSettingsActivity.java:46)
        at android.app.Activity.performCreate(Activity.java:6684)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1119)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2637)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2751) 
        at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4556) 
        at android.app.ActivityThread.-wrap19(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1502) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:154) 
        at android.app.ActivityThread.main(ActivityThread.java:6186) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
```

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -139,6 +139,11 @@ dependencies {
     implementation "com.android.support:gridlayout-v7:$supportVersion"
     implementation "com.android.support:recyclerview-v7:$supportVersion"
     compileOnly 'com.google.android.wearable:wearable:2.2.0'
+
+    // ViewModel and LiveData
+    implementation "android.arch.lifecycle:extensions:$lifecycle_version"
+    annotationProcessor "android.arch.lifecycle:compiler:$lifecycle_version"
+
     implementation "org.apache.commons:commons-lang3:$commonslangVersion"
     implementation("org.shredzone.flattr4j:flattr4j-core:$flattr4jVersion") {
         exclude group: "org.json", module: "json"

--- a/app/src/main/java/de/danoeh/antennapod/activity/FeedSettingsActivity.java
+++ b/app/src/main/java/de/danoeh/antennapod/activity/FeedSettingsActivity.java
@@ -1,5 +1,6 @@
 package de.danoeh.antennapod.activity;
 
+import android.arch.lifecycle.ViewModelProviders;
 import android.graphics.LightingColorFilter;
 import android.os.Bundle;
 import android.support.v4.app.FragmentManager;
@@ -7,7 +8,6 @@ import android.support.v4.app.FragmentTransaction;
 import android.support.v7.app.AppCompatActivity;
 import android.text.TextUtils;
 import android.util.Log;
-
 import android.view.MenuItem;
 import android.view.View;
 import android.widget.ImageView;
@@ -19,10 +19,8 @@ import de.danoeh.antennapod.core.feed.Feed;
 import de.danoeh.antennapod.core.glide.ApGlideSettings;
 import de.danoeh.antennapod.core.glide.FastBlurTransformation;
 import de.danoeh.antennapod.core.preferences.UserPreferences;
-import de.danoeh.antennapod.core.storage.DBReader;
 import de.danoeh.antennapod.fragment.FeedSettingsFragment;
-import io.reactivex.Maybe;
-import io.reactivex.MaybeOnSubscribe;
+import de.danoeh.antennapod.viewmodel.FeedSettingsViewModel;
 import io.reactivex.android.schedulers.AndroidSchedulers;
 import io.reactivex.disposables.Disposable;
 import io.reactivex.schedulers.Schedulers;
@@ -46,7 +44,6 @@ public class FeedSettingsActivity extends AppCompatActivity {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.feedsettings);
         getSupportActionBar().setDisplayHomeAsUpEnabled(true);
-        long feedId = getIntent().getLongExtra(EXTRA_FEED_ID, -1);
 
         imgvCover = findViewById(R.id.imgvCover);
         txtvTitle = findViewById(R.id.txtvTitle);
@@ -57,14 +54,8 @@ public class FeedSettingsActivity extends AppCompatActivity {
         // https://github.com/bumptech/glide/issues/529
         imgvBackground.setColorFilter(new LightingColorFilter(0xff828282, 0x000000));
 
-        disposable = Maybe.create((MaybeOnSubscribe<Feed>) emitter -> {
-                    Feed feed = DBReader.getFeed(feedId);
-                    if (feed != null) {
-                        emitter.onSuccess(feed);
-                    } else {
-                        emitter.onComplete();
-                    }
-                })
+        long feedId = getIntent().getLongExtra(EXTRA_FEED_ID, -1);
+        disposable = ViewModelProviders.of(this).get(FeedSettingsViewModel.class).getFeed(feedId)
                 .subscribeOn(Schedulers.io())
                 .observeOn(AndroidSchedulers.mainThread())
                 .subscribe(result -> {
@@ -89,10 +80,6 @@ public class FeedSettingsActivity extends AppCompatActivity {
                 fragmentManager.beginTransaction();
         fragmentTransaction.replace(R.id.settings_fragment_container, fragment);
         fragmentTransaction.commit();
-    }
-
-    public Feed getFeed() {
-        return feed;
     }
 
     private void showHeader() {

--- a/app/src/main/java/de/danoeh/antennapod/fragment/FeedSettingsFragment.java
+++ b/app/src/main/java/de/danoeh/antennapod/fragment/FeedSettingsFragment.java
@@ -1,5 +1,6 @@
 package de.danoeh.antennapod.fragment;
 
+import android.arch.lifecycle.ViewModelProviders;
 import android.content.Context;
 import android.content.DialogInterface;
 import android.os.Bundle;
@@ -7,7 +8,6 @@ import android.support.v14.preference.SwitchPreference;
 import android.support.v7.preference.ListPreference;
 import android.support.v7.preference.PreferenceFragmentCompat;
 import de.danoeh.antennapod.R;
-import de.danoeh.antennapod.activity.FeedSettingsActivity;
 import de.danoeh.antennapod.core.dialog.ConfirmationDialog;
 import de.danoeh.antennapod.core.feed.Feed;
 import de.danoeh.antennapod.core.feed.FeedFilter;
@@ -16,6 +16,9 @@ import de.danoeh.antennapod.core.preferences.UserPreferences;
 import de.danoeh.antennapod.core.storage.DBWriter;
 import de.danoeh.antennapod.dialog.AuthenticationDialog;
 import de.danoeh.antennapod.dialog.EpisodeFilterDialog;
+import de.danoeh.antennapod.viewmodel.FeedSettingsViewModel;
+
+import static de.danoeh.antennapod.activity.FeedSettingsActivity.EXTRA_FEED_ID;
 
 public class FeedSettingsFragment extends PreferenceFragmentCompat {
     private static final CharSequence PREF_EPISODE_FILTER = "episodeFilter";
@@ -26,17 +29,21 @@ public class FeedSettingsFragment extends PreferenceFragmentCompat {
     public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
         addPreferencesFromResource(R.xml.feed_settings);
 
-        feed = ((FeedSettingsActivity) getActivity()).getFeed();
-        feedPreferences = feed.getPreferences();
+        long feedId = getArguments().getLong(EXTRA_FEED_ID);
+        ViewModelProviders.of(getActivity()).get(FeedSettingsViewModel.class).getFeed(feedId)
+                .subscribe(result -> {
+                    feed = result;
+                    feedPreferences = feed.getPreferences();
 
-        setupAutoDownloadPreference();
-        setupKeepUpdatedPreference();
-        setupAutoDeletePreference();
-        setupAuthentificationPreference();
-        setupEpisodeFilterPreference();
+                    setupAutoDownloadPreference();
+                    setupKeepUpdatedPreference();
+                    setupAutoDeletePreference();
+                    setupAuthentificationPreference();
+                    setupEpisodeFilterPreference();
 
-        updateAutoDeleteSummary();
-        updateAutoDownloadEnabled();
+                    updateAutoDeleteSummary();
+                    updateAutoDownloadEnabled();
+                }).dispose();
     }
 
     private void setupEpisodeFilterPreference() {

--- a/app/src/main/java/de/danoeh/antennapod/viewmodel/FeedSettingsViewModel.java
+++ b/app/src/main/java/de/danoeh/antennapod/viewmodel/FeedSettingsViewModel.java
@@ -1,0 +1,30 @@
+package de.danoeh.antennapod.viewmodel;
+
+import android.arch.lifecycle.ViewModel;
+import de.danoeh.antennapod.core.feed.Feed;
+import de.danoeh.antennapod.core.storage.DBReader;
+import io.reactivex.Maybe;
+
+public class FeedSettingsViewModel extends ViewModel {
+    private Feed feed;
+
+    public Maybe<Feed> getFeed(long feedId) {
+        if (feed == null) {
+            return loadFeed(feedId);
+        } else {
+            return Maybe.just(feed);
+        }
+    }
+
+    private Maybe<Feed> loadFeed(long feedId) {
+        return Maybe.create(emitter -> {
+            Feed feed = DBReader.getFeed(feedId);
+            if (feed != null) {
+                this.feed = feed;
+                emitter.onSuccess(feed);
+            } else {
+                emitter.onComplete();
+            }
+        });
+    }
+}

--- a/build.gradle
+++ b/build.gradle
@@ -45,6 +45,7 @@ project.ext {
     targetSdkVersion = 26
 
     supportVersion = "27.1.1"
+    lifecycle_version = "1.1.1"
     workManagerVersion = "1.0.1"
     awaitilityVersion = "3.1.2"
     commonsioVersion = "2.5"

